modules/gdscript/gdscript_parser.cpp

```diff
@@ -4157,6 +4162,39 @@ bool GDScriptParser::onready_annotation(AnnotationNode *p_annotation, Node *p_ta
 	return true;
 }
 
+bool GDScriptParser::tool_button_annotation(AnnotationNode *p_annotation, Node *p_node, ClassNode *p_class) {
+#ifdef TOOLS_ENABLED
+	AnnotationNode *annotation = const_cast<AnnotationNode *>(p_annotation);
+	FunctionNode *func = static_cast<FunctionNode *>(p_node);
+
+	if (!is_tool()) {
+		push_error("Tool buttons can only be used in tool scripts (add `@tool` to the top of the script).", p_annotation);
+		return false;
+	}
+
+	annotation->export_info.name = "__tool_button_" + func->identifier->name;
+	annotation->export_info.type = Variant::Type::CALLABLE;
+	annotation->export_info.usage = PROPERTY_USAGE_EDITOR | PROPERTY_USAGE_INTERNAL;
+
+	if (func->parameters.size() >= 1 && func->parameters[0]->datatype_specifier != nullptr && func->parameters[0]->datatype_specifier->type_chain[0]->name != "EditorUndoRedoManager") {
```

Comment:
Looks unsafe (no `nullptr` and bound checks).

Reply:
Perhaps I am missing something, can you be explicit where the problem is?

Reply:
I've had a good look at the usages of `func->parameters` and `type_chain` and their elements, nowhere else is doing this level of `nullptr` checking. Can you be explicit about where nulls are permitted by the system and which require checking?

```cpp
if (func->parameters.size() >= 1) {
  auto p0 = func->parameters[0];
  auto p0_dt = p0 ? p0->datatype_specifier : nullptr;
  auto t0 = p0_dt && p0_dt->type_chain.size() >= 1 ? p0_dt->type_chain[0] : nullptr;

  if (t0 && t0->name != "EditorUndoRedoManager") {
    push_error("Tool button methods should have their first argument typed as an EditorUndoRedoManager.", func);
    return false;
  }
}
```

Reply:
This isn't a suggested change; I just want to lay out how ambiguous the ask is and that no other methods are going to these lengths.

Reply:
Instead of `type_chain` you can use `DataType`, see the review below.